extglob.go

```diff
@@ -1,25 +1,179 @@
 package main
 
 import (
-	"fmt"
-	"os"
 	"regexp"
+	"regexp/syntax"
 	"strings"
 )
 
-var (
-	parensExpr   = regexp.MustCompile("([^\\\\])\\(([^)]+)\\)")
-	questionExpr = regexp.MustCompile("\\?([^(])")
-)
-
 func CompileExtGlob(extglob string) (*regexp.Regexp, error) {
-	tmp := extglob
-	tmp = strings.Replace(tmp, ".", "\\.", -1)
-	tmp = strings.Replace(tmp, "**", ".🐷", -1)
-	tmp = strings.Replace(tmp, "*", fmt.Sprintf("[^%c]*", os.PathSeparator), -1)
-	tmp = questionExpr.ReplaceAllString(tmp, fmt.Sprintf("[^%c]$1", os.PathSeparator))
-	tmp = parensExpr.ReplaceAllString(tmp, "($2)$1")
-	tmp = strings.Replace(tmp, ")@", ")", -1)
-	tmp = strings.Replace(tmp, ".🐷", ".*", -1)
-	return regexp.Compile("^" + tmp + "$")
+	ctx := globctx{glob: extglob}
+	ctx.compileGlobstarPrefix()
+
+	if err := ctx.compileExpression(); err != nil {
+		return nil, err
+	}
+
+	return regexp.Compile("^" + string(ctx.regexp) + "$")
+}
+
+type globctx struct {
+	glob       string
+	regexp     []byte
+	pos, depth int
+}
+
+func (c *globctx) compileExpression() error {
+	for c.pos < len(c.glob) {
+		switch curr := c.glob[c.pos]; curr {
+		case '\\':
+			if err := c.compileEscapeSequence(); err != nil {
+				return err
+			}
+		case '*':
+			if err := c.compileSubExpression("(?:", ")*", "[^/]*"); err != nil {
+				return err
+			}
+		case '?':
+			if err := c.compileSubExpression("(?:", ")?", "[^/]"); err != nil {
+				return err
+			}
+		case '+':
+			if err := c.compileSubExpression("(?:", ")+", "\\+"); err != nil {
+				return err
+			}
+		case '@':
+			if err := c.compileSubExpression("(?:", ")", "\\@"); err != nil {
+				return err
+			}
+		case '!':
+			if err := c.compileSubExpression("(?~", ")", "\\!"); err != nil {
```

Comment:
Negative matches are not actually supported. Go regexp supports neither negative lookahead `(?!)`, nor the Ruby's absence operator `(?~)` that we'd need to implement this. This returns an `ErrInvalidPerlOp`.